### ISSUE TITLE
Selection observer should call render

### ISSUE THREAD

**Issue by pjasiun** [ckeditor/ckeditor5-engine#796]
The idea is that observers not only listen on DOM changes and fire view events but also take care about re-rendering DOM if it needs to be rendered. Especially, if the model did not fire `change` nor `changesDone` events it's observers role to bring DOM to the same state as view is, by calling `render()`.

Note that it's renderer who should compare DOM and view and do nothing if there is no good reason to render anything. Observers should be free to call render whenever the feel that something may need the update.

`render` was removed from the `SelectionObserver` by this commit: https://github.com/ckeditor/ckeditor5-engine/commit/f4c0d25398bf9a85376e2f24f32c26cff188e16e

Because of it 'noselection' manual test does not work: http://localhost:8125/ckeditor5-engine/tests/view/manual/noselection.html

It's also connected with other issues: https://github.com/ckeditor/ckeditor5-engine/issues/795

**Comment by Reinmar** [comment]
I needed to change a code of http://localhost:8125/ckeditor5-engine/tests/view/manual/inline-filler.html a bit to make the test pass now.

Also, the http://localhost:8125/ckeditor5-engine/tests/view/manual/immutable.html test doesn't work now due to this.

**Comment by szymonkups** [comment]
I am working on `UIElement` rendering now. All DOM selection that lands inside `UIElement` should be converted by `DOMConverter` to selection before/after `UIElement`. This means that each selection change inside `UIElement` is converted to the same selection. When `SelectionObserver` is not calling `render()`, we end up with "dirty" selection in DOM having correct selection in view and model.

**Comment by ma2ciek** [comment]
In https://github.com/ckeditor/ckeditor5/issues/676 and https://github.com/ckeditor/ckeditor5-engine/issues/1157 we face a problem, that the `focus` and `selectionchange` events fire with some delay between them, which results in messing up the editor's selection (see https://github.com/ckeditor/ckeditor5/issues/676#issuecomment-349307800). We need to make the selection depend more on the `selectionchange` event, less on the `focus` event or make it somehow smarter.

> Observers should be free to call render whenever the feel that something may need the update.

For now, calling `render()` between `selectionchange` and `focus` events results in errors.

**Comment by ma2ciek** [comment]
#1153 and #1155 are also connected with that topic.

**Comment by ma2ciek** [comment]
And, I omitted one important thing, rendering in both `FocusObserver` and `SelectionObserver` won't fix the problem because sync rendering in the `FocusObserver` part crashes the model selection and the focus (DOM selection at this stage is not updated and is taken from the wrong place). Eventually, SelectionObserver might later try to fix it, but it doesn't look like good. it would create a gap between two fired events where the model is invalid.

**Comment by Reinmar** [comment]
Yep, the only right place to render is after selection change. Which makes sense as long as selection change is always fired after focus. This needs to be verified. So far we've seen that focus always triggers selection change, but this is the kind of things you can never be sure, so perhaps some safety mechanism could be found.
